lisa/trace.py

```diff
@@ -1115,11 +1126,27 @@ def __call__(self, f):
         else:
             checker = AndTraceEventChecker([self, used_events])
 
-        @wraps(f)
-        def wrapper(self, *args, **kwargs):
-            available_events = set(self.trace.available_events)
-            checker.check_events(available_events)
-            return f(self, *args, **kwargs)
+        sig = inspect.signature(f)
+        if sig.parameters:
+            @wraps(f)
+            def wrapper(self, *args, **kwargs):
+                try:
+                    trace = self.trace
+                # If there is no "trace" attribute, silently skip the check. This
+                # allows using the decorator for documentation and chaining purpose
+                # without having an actual trace to work on.
+                except AttributeError:
+                    pass
+                else:
+                    available_events = set(trace.available_events)
+                    checker.check_events(available_events)
+
+                return f(self, *args, **kwargs)
+        # If the decorated object takes no parameters, we cannot check anything
+        else:
+            @wraps(f)
+            def wrapper(*args, **kwargs):
```

Comment:
Can't we just `return f` here?

Reply:
Oh no we can still add the `used_events` - don't mind me!

Reply:
we could return `f` and just stick the events on it, but I'd rather keep the same number of decorator levels in all paths, to avoid later surprises.